cgo: `go link` detects all linker flags as unsupported when using custom cc toolchain

<!-- Please answer these questions before submitting your issue. Thanks! -->

### What version of rules_go are you using?

master sync'd to latest

```
mike.danese@ip-10-110-30-195:~/rules_go$ git show
commit 36e04e9f1e9b56c865332edfb49e1620beaedf2c
```

### What version of Bazel are you using?


```
mike.danese@ip-10-110-30-195:~/rules_go$ bazel version
Bazelisk version: development
Build label: 7.1.0
Build target: @@//src/main/java/com/google/devtools/build/lib/bazel:BazelServer
Build time: Mon Mar 11 17:55:31 2024 (1710179731)
Build timestamp: 1710179731
Build timestamp as int: 1710179731
mike.danese@ip-10-110-30-195:~/rules_go$ git show
commit 36e04e9f1e9b56c865332edfb49e1620beaedf2c
```

### Does this issue reproduce with the latest releases of all the above?



### What operating system and processor architecture are you using?

Linux

### What did you see instead?

I have a custom cc toolchain based on LLVM. This toolchain sets a sysroot link option, e.g. `--sysroot=external/sysroot-linux-x86_64-gnu7-llvm-1`.

When a cgo library is generated, the cgo tool converts this LD_FLAG value into an absolute path here: https://cs.opensource.google/go/go/+/master:src/cmd/cgo/main.go;l=353-357;drc=bdccd923e914ab61d77a8f23a3329cf1d5aaa7c1. For example, my sysroot ldflag might be converted to `/home/mike.danese/.cache/bazel/_bazel_mike.danese/71e455921cc87b7db5489570176e2cff/sandbox/linux-sandbox/400/execroot/universe/external/sysroot-linux-x86_64-gnu7-llvm-16`. These LD_FLAGs get backed into the object.

Now when it's time to link, `go tool link` probes the linker (in `linkerFlagSupported`) to determine if e.g. `-no-pie` should be configured: https://cs.opensource.google/go/go/+/master:src/cmd/link/internal/ld/lib.go;l=1833-1834;drc=2ab9218c86ed625362df5060f64fcd59398a76f3.  It does this by compiling a trivial program. To construct the compile command, the ldflags from the cgo object are appended to the extdld flags: https://cs.opensource.google/go/go/+/master:src/cmd/link/internal/ld/lib.go;l=2079-2080;drc=2ab9218c86ed625362df5060f64fcd59398a76f3. This results in a command like this:

```
external/llvm-toolchain-16-0-4-linux-x86_64/bin/clang \
  -m64 -fuse-ld=lld -Wl,--build-id=md5 -Wl,--hash-style=gnu -Wl,-z,relro,-z,now -Wl,-z,separate-code \
  --sysroot=external/sysroot-linux-x86_64-gnu7-llvm-16 -pthread -fuse-ld=lld \
  -Wl,--build-id=md5 -Wl,--hash-style=gnu -Wl,-z,relro,-z,now -Wl,-z,separate-code \
  --sysroot=/home/mike.danese/.cache/bazel/_bazel_mike.danese/71e455921cc87b7db5489570176e2cff/sandbox/linux-sandbox/423/execroot/universe/external/sysroot-linux-x86_64-gnu7-llvm-16 \
  -pthread -o /tmp/go-link-606451175/a.out -no-pie /tmp/go-link-606451175/trivial.c
```

`sysroot` arg is passed twice, once from --extld flags and once from the go_ldflags attribute in the cgo object so the second `--sysroot` wins. This is a problem because the second sysroot references a path in the no longer existant execution sandbox where the cgo object was compiled. Thus, the probe fails with a nonsensical error:

```
ld.lld: error: cannot open crt1.o: No such file or directory
ld.lld: error: cannot open crti.o: No such file or directory
ld.lld: error: cannot open crtbegin.o: No such file or directory
ld.lld: error: unable to find library -lgcc
ld.lld: error: unable to find library -lgcc_s
ld.lld: error: unable to find library -lpthread
ld.lld: error: unable to find library -lc
ld.lld: error: unable to find library -lgcc
ld.lld: error: unable to find library -lgcc_s
ld.lld: error: cannot open crtend.o: No such file or directory
ld.lld: error: cannot open crtn.o: No such file or directory
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

This error is swallowed silently, and `go link` assumes that the linker flag is unsupported. In the case of `-no-pie`, this causes rules_go to attempt to link PIE cgo objects into go binaries without relro (e.g. in normal exe buildmode on linux). This causes hard to debug runtime errors such as:

```
runtime: pcHeader: magic= 0xfffffff1 pad1= 0 pad2= 0 minLC= 1 ptrSize= 8 pcHeader.textStart= 0x1be100 text= 0x557164240100 pluginpath=
fatal error: invalid function symbol table
runtime: panic before malloc heap initialized

runtime stack:
runtime.throw({0x5571640f265a?, 0x0?})
        GOROOT/src/runtime/panic.go:1077 +0x5c fp=0x7fff932345a8 sp=0x7fff93234578 pc=0x55716427741c
runtime.moduledataverify1(0x8?)
        GOROOT/src/runtime/symtab.go:533 +0x816 fp=0x7fff932346c8 sp=0x7fff932345a8 pc=0x5571642958f6
runtime.moduledataverify(...)
        GOROOT/src/runtime/symtab.go:519
runtime.schedinit()
        GOROOT/src/runtime/proc.go:726 +0x4c fp=0x7fff93234710 sp=0x7fff932346c8 pc=0x55716427af6c
runtime.rt0_go()
        src/runtime/asm_amd64.s:349 +0x11c fp=0x7fff93234718 sp=0x7fff93234710 pc=0x5571642aa17c
```

There are ~10 other calls to `linkerFlagSupported` and none of them are correctly detecting flag support.

Ah ya, misread that. Thanks.

So this is actually happening in rules_go. The flags are getting absolute here:

https://github.com/bazelbuild/rules_go/blob/aeb83e878033ef357642c87122e193df44da03fe/go/tools/builders/stdlib.go#L161-L165

So we have a --sysroot from a custom cc toolchain. When building the stdlib with `go install`, we make all CGO_LDFLAGS absolute, presumably because `go install` changes directories. If we don't, then we get an error like:

```
_cgo_export.c:3:10: fatal error: 'stdlib.h' file not found
```

--- a/go/tools/builders/BUILD.bazel
+++ b/go/tools/builders/BUILD.bazel
@@ -57,6 +57,16 @@ go_test(
     ],
 )
 
+go_test(
+    name = "env_test",
+    size = "small",
+    srcs = [
+        "env.go",
+        "env_test.go",
+        "flags.go",
+    ],
+)
+
 filegroup(
     name = "builder_srcs",
     srcs = [

--- a/go/tools/builders/builder.go
+++ b/go/tools/builders/builder.go
@@ -32,16 +32,27 @@ func main() {
 	if err != nil {
 		log.Fatal(err)
 	}
-	if len(args) == 0 {
+
+	verb := verbFromName(os.Args[0])
+	if verb == "" && len(args) == 0 {
 		log.Fatalf("usage: %s verb options...", os.Args[0])
 	}
-	verb, rest := args[0], args[1:]
+
+	var rest []string
+	if verb == "" {
+		verb, rest = args[0], args[1:]
+	} else {
+		rest = args
+	}
 
 	var action func(args []string) error
 	switch verb {
-	case "compilepkg": action = compilePkg
-	case "nogo": action = nogo
-	case "nogovalidation": action = nogoValidation
+	case "compilepkg":
+		action = compilePkg
+	case "nogo":
+		action = nogo
+	case "nogovalidation":
+		action = nogoValidation
 	case "filterbuildid":
 		action = filterBuildID
 	case "gentestmain":

--- a/go/tools/builders/cc.go
+++ b/go/tools/builders/cc.go
@@ -13,12 +13,13 @@ import (
 func cc(args []string) error {
 	cc := os.Getenv("GO_CC")
 	if cc == "" {
-		errors.New("GO_CC environment variable not set")
+		return errors.New("GO_CC environment variable not set")
 	}
 	ccroot := os.Getenv("GO_CC_ROOT")
 	if ccroot == "" {
-		errors.New("GO_CC_ROOT environment variable not set")
+		return errors.New("GO_CC_ROOT environment variable not set")
 	}
+
 	normalized := []string{cc}
 	normalized = append(normalized, args...)
 	transformArgs(normalized, cgoAbsEnvFlags, func(s string) string {
@@ -26,10 +27,10 @@ func cc(args []string) error {
 			trimmed := strings.TrimPrefix(s, cgoAbsPlaceholder)
 			abspath := filepath.Join(ccroot, trimmed)
 			if _, err := os.Stat(abspath); err == nil {
-                                // Only return the abspath if it exists, otherwise it
-                                // means that either it won't have any effect or the original
-                                // value was not a relpath (e.g. a path with a XCODE placehold from
-                                // macos cc_wrapper)
+				// Only return the abspath if it exists, otherwise it
+				// means that either it won't have any effect or the original
+				// value was not a relpath (e.g. a path with a XCODE placehold from
+				// macos cc_wrapper)
 				return abspath
 			}
 			return trimmed

--- a/go/tools/builders/env.go
+++ b/go/tools/builders/env.go
@@ -25,6 +25,7 @@ import (
 	"os"
 	"os/exec"
 	"path/filepath"
+	"regexp"
 	"runtime"
 	"strconv"
 	"strings"
@@ -160,6 +161,86 @@ func (e *env) runCommandToFile(out, err io.Writer, args []string) error {
 	cmd.Stdout = out
 	cmd.Stderr = err
 	return runAndLogCommand(cmd, e.verbose)
+}
+
+func symlinkBuilderCC() (func(), string, error) {
+	absPath, err := filepath.Abs(os.Args[0])
+	if err != nil {
+		return nil, "", err
+	}
+
+	dirname, err := os.MkdirTemp("", "")
+	if err != nil {
+		return nil, "", err
+	}
+
+	// re-use the file extension if there is one (mostly for .exe)
+	builderPath := filepath.Join(dirname, "builder-cc"+filepath.Ext(os.Args[0]))
+
+	if err := os.Symlink(absPath, builderPath); err != nil {
+		_ = os.RemoveAll(dirname)
+		return nil, "", err
+	}
+
+	return func() { _ = os.RemoveAll(dirname) }, builderPath, nil
+}
+
+var symlinkBuilderName = regexp.MustCompile(`.*builder-(\w+)(\.exe)?$`)
+
+// verbFromName parses the builder verb from the builder name so
+// that the builder+verb can be invoked in a way that allows the verb
+// to be part of the program name, rather than requring it to be in the
+// arg list. this makes it easier to set a program like `builder cc` as
+// an environment variable in a way that is more compatible
+func verbFromName(arg0 string) string {
+	matches := symlinkBuilderName.FindAllStringSubmatch(arg0, 1)
+	if len(matches) != 1 {
+		return ""
+	}
+
+	if len(matches[0]) < 2 {
+		return ""
+	}
+
+	return matches[0][1]
+}
+
+// absCCLinker sets the target of the `-extld` to a name which `verbFromName`
+// can extract correctly, allowing go tool link to be able to use the `builder cc` wrapper
+func absCCLinker(argList []string) (func(), error) {
+	extldIndex := -1
+	for i, arg := range argList {
+		if arg == "-extld" && i+1 < len(argList) {
+			extldIndex = i + 1
+			break
+		}
+	}
+	if extldIndex < 0 {
+		// if we don't find extld just return
+		// a noop cleanup function
+		return func() {}, nil
+	}
+
+	cleanup, buildercc, err := symlinkBuilderCC()
+	if err != nil {
+		return nil, err
+	}
+
+	err = os.Setenv("GO_CC", argList[extldIndex])
+	if err != nil {
+		cleanup()
+		return nil, err
+	}
+
+	err = os.Setenv("GO_CC_ROOT", abs("."))
+	if err != nil {
+		cleanup()
+		return nil, err
+	}
+
+	argList[extldIndex] = buildercc
+
+	return cleanup, nil
 }
 
 // absCCCompiler modifies CGO flags to workaround relative paths.

--- a/go/tools/builders/env_test.go
+++ b/go/tools/builders/env_test.go
@@ -1,0 +1,26 @@
+//go:build !windows
+
+package main
+
+import "testing"
+
+func TestVerbFromName(t *testing.T) {
+	testCases := []struct {
+		name string
+		verb string
+	}{
+		{"/a/b/c/d/builder", ""},
+		{"builder", ""},
+		{"/a/b/c/d/builder-cc", "cc"},
+		{"builder-ld", "ld"},
+		{"c:\\builder\\builder.exe", ""},
+		{"c:\\builder with spaces\\builder-cc.exe", "cc"},
+	}
+
+	for _, tc := range testCases {
+		result := verbFromName(tc.name)
+		if result != tc.verb {
+			t.Fatalf("retrieved invalid verb %q from name %q", result, tc.name)
+		}
+	}
+}

--- a/go/tools/builders/link.go
+++ b/go/tools/builders/link.go
@@ -141,6 +141,19 @@ func link(args []string) error {
 	}
 	goargs = append(goargs, "-o", *outFile)
 
+	// substitute `builder cc` for the linker with a symlink to builder called `builder-cc`.
+	// unfortunately we can't just set an environment variable to `builder cc` because
+	// in `go tool link` the `linkerFlagSupported` [1][2] call sites used to determine
+	// if a linker supports various flags all appear to use the first arg after splitting
+	// so the `cc` would be left off of `builder cc`
+	//
+	//    [1]: https://cs.opensource.google/go/go/+/ad7f736d8f51ea03166b698256385c869968ae3e:src/cmd/link/internal/ld/lib.go;l=1739
+	//    [2]: https://cs.opensource.google/go/go/+/master:src/cmd/link/internal/ld/lib.go;drc=c6531fae589cf3f9475f3567a5beffb4336fe1d6;l=1429?q=linkerFlagSupported&ss=go%2Fgo
+	linkerCleanup, err := absCCLinker(toolArgs)
+	if err != nil {
+		return err
+	}
+	defer linkerCleanup()
 	// add in the unprocess pass through options
 	goargs = append(goargs, toolArgs...)
 	goargs = append(goargs, *main)

--- a/tests/core/cgo/BUILD.bazel
+++ b/tests/core/cgo/BUILD.bazel
@@ -400,6 +400,11 @@ go_test(
     cgo = True,
 )
 
+go_bazel_test(
+    name = "cgo_abs_paths_test",
+    srcs = ["cgo_abs_paths_test.go"],
+)
+
 cc_library(
     name = "cgo_link_dep",
     srcs = ["cgo_link_dep.c"],

--- a/tests/core/cgo/cgo_abs_paths_test.go
+++ b/tests/core/cgo/cgo_abs_paths_test.go
@@ -1,0 +1,82 @@
+package cgo_abs_paths_test
+
+import (
+	"strings"
+	"testing"
+
+	"github.com/bazelbuild/rules_go/go/tools/bazel_testing"
+)
+
+func TestMain(m *testing.M) {
+	bazel_testing.TestMain(m, bazel_testing.Args{
+		Main: `
+-- main.go --
+package main
+
+/*
+#include <stdio.h>
+
+void hello() {
+	printf("Hello, world!\n");
+}
+*/
+import "C"
+
+func main() {
+	C.hello()
+}
+-- BUILD.bazel --
+load("@io_bazel_rules_go//go:def.bzl", "go_test", "go_library")
+
+go_library(
+    name = "example_lib",
+    srcs = ["main.go"],
+	cgo = True,
+    importpath = "example.com/cmd/example",
+)
+
+go_test(
+    name = "example",
+    embed = [":example_lib"],
+	pure = "off",
+)
+`,
+		SetUp: func() error {
+			// we have to force a bazel clean because this test needs to
+			// fully execute every time so that we can actually scan the output
+			return bazel_testing.RunBazel("clean")
+		},
+	})
+}
+
+func TestCgoAbsolutePaths(t *testing.T) {
+	_, stderrBytes, err := bazel_testing.BazelOutputWithInput(nil,
+		"build",
+		"--linkopt=-Lan/imaginary/lib",
+		"--linkopt=-v",
+		"--copt=-Ian/imaginary/include",
+		"--copt=-v",
+		"-s",
+		"//:example",
+	)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	stderr := string(stderrBytes)
+
+	if strings.Contains(stderr, "__GO_BAZEL_CC_PLACEHOLDER__") {
+		t.Log(stderr)
+		t.Fatal("Found absolute path placeholder string in go linker command output")
+	}
+
+	if !strings.Contains(stderr, "-Lan/imaginary/lib") {
+		t.Log(stderr)
+		t.Fatal("Could not find --linkopt in go linker command output")
+	}
+
+	if !strings.Contains(stderr, "-Ian/imaginary/include") {
+		t.Log(stderr)
+		t.Fatal("Could not find --copt in go compiler command output")
+	}
+}

